setup.py

```diff
@@ -234,7 +234,8 @@ def get_build_ext_override():
             from pythran.dist import PythranBuildExt
             BaseBuildExt = PythranBuildExt[npy_build_ext]
             # Version check - a too old Pythran will give problems
-            if LooseVersion(pythran.__version__) < LooseVersion('0.9.12'):
+            importlib.import_module('scipy/_lib/_pep440')
```

Comment:
this is gruesome, and I'm not sure we are "allowed" to do this, but we cannot import from within `scipy` directly from within `setup.py`  as far as I can tell (it dumps a bunch of chicken-and-egg type init errors)

Reply:
There's a line`from scipy._build_utils.system_info import get_info, NotFoundError` inside `configuration()`, so it's not completely forbidden. But yeah, if it doesn't work then this workaround is fine.

Reply:
Double checked, this should be fine.

Reply:
I get build problems using the latest main (as of now):
```
Traceback (most recent call last):
  File "/home/nicpa/codes/scipy/setup.py", line 530, in <module>
    setup_package()
  File "/home/nicpa/codes/scipy/setup.py", line 514, in setup_package
    generate_cython()
  File "/home/nicpa/codes/scipy/setup.py", line 255, in generate_cython
    if _pep440.parse(pip.__version__) < _pep440.Version('18.0.0'):
NameError: name '_pep440' is not defined
```
this seems related to this PR?

Reply:
It seems this line is missing in line 255 as well...

I can make a PR if you want?

Reply:
yes please!

Reply:
if you could add the build command you are using, that'd be helpful. because all our CI is passing, so unclear to me why it's failing for you.

Reply:
I found out that I had issued the commands wrong, manually using `build_ext` etc. I had an old script I was using thoughtlessly.
It seems that if users manually run `python3 setup.py` this occurs? (Not completely sure/tested)

I went to the `pip` installation procedure and now all works fine.
Thanks!

Reply:
No, `python setup.py xxx` should work fine, I use it all the time. `pip` is not a real replacement as a developer interface unfortunately.

Reply:
I use `pip --no-build-isolation` which makes it straightforwardly usable as developer interface. Forces the use of installed packages.

Ok, I will try and see if I can trigger it again.

Reply:
Hmm, now I can't reproduce it... 
I tried checking out the commit before my PR, and reran with `setup.py install` like my old script... But I can't get the same problem.

Perhaps the submodules did something here... I am at a loss with this... Sorry...

Reply:
I wish it were that easy. It will still copy to a tmpdir (slow rebuilds), and hide the build log. I believe the closest equivalent to `python setup.py develop` is `pip install -v --no-build-isolation --no-use-pep517 -e .`. Which is a bit sad:(

I'm still pondering on how we get an optimal CLI for the Meson build, because `pip` is unlikely to be the right answer.

Reply:
> I wish it were that easy. It will still copy to a tmpdir (slow rebuilds), and hide the build log. I believe the closest equivalent to `python setup.py develop` is `pip install -v --no-build-isolation --no-use-pep517 -e .`. Which is a bit sad:(

Hmm... I see... Yeah, this is not really the *pip way*...

> I'm still pondering on how we get an optimal CLI for the Meson build, because `pip` is unlikely to be the right answer.

And I have been following this closely, I am very much interested in seeing what Meson can do for Python builds. But I kind-of feel that this requires a complete restructure of the build infrastructure of Python if pip can't be *patched*. And then we go again with build-systems.

Perhaps the infrastructure should be in _meson_ extensions instead?

Just some random thoughts.

Reply:
> Perhaps the infrastructure should be in _meson_ extensions instead?

Yes, probably. I've also looked at Nox, which does a lot of things right but it seems to have the annoying habit of coupling everything it does with virtualenvs, and it's not clear on how to get the good parts of Nox without that.

Reply:
I don't know nox, but just by having a look at its documentation page, it doesn't seem like it will fix anything?

To me it is important to be able to build against already installed packages. And many of these venv stuff needs customization flags for running custom builds... Hmm....

Reply:
What I like from a tool like Nox (in principle) is that it'd give building blocks for things that we now do completely by hand in `runtests.py` and `dev.py`. We shouldn't need ~600 LoC to implement our own developer CLI.

> To me it is important to be able to build against already installed packages.

Yes, this should be the default for any sane interface. You set up a build env once (which can be expensive), and then you build/test/benchmark over and over.

Reply:
Yes! :)

There's a reason there are so many alternatives, a stable *all-across* solution is not easy :-D